numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -328,28 +328,31 @@ gentype_nonzero_number(PyObject *m1)
 static PyObject *
 gentype_str(PyObject *self)
 {
-    PyObject *arr, *ret = NULL;
+    PyObject  *item, *item_str;
 
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        ret = PyObject_Str((PyObject *)arr);
-        Py_DECREF(arr);
+    item = gentype_generic_method(self, NULL, NULL, "item");
+    if (item == NULL) {
+        return NULL;
     }
-    return ret;
+
+    item_str = PyObject_Str(item);
```

Comment:
As I think I said before, this isn't safe, since `item` can return a copy of `self`, causing infinite recursion.

What actually falls back to this case anyway?
Seems to me that it would be better to just inherit `object.__repr__`, making if obvious that custom dtype users have work to do.